R/human_infection.R

```diff
@@ -48,91 +51,124 @@ simulate_infection <- function(
+    # get vector of individuals who have received their 3rd dose
+    vaccinated <- vaccine_times > -1
+    pev_profile <- pev_profile[vaccinated]
+    if (length(vaccinated) > 0) {
+      antibodies <- calculate_pev_antibodies(
+        timestep - vaccine_times[vaccinated],
+        exp(sample_pev_param(pev_profile, parameters$pev_profiles, 'cs')),
+        invlogit(sample_pev_param(pev_profile, parameters$pev_profiles, 'rho')),
+        exp(sample_pev_param(pev_profile, parameters$pev_profiles, 'ds')),
+        exp(sample_pev_param(pev_profile, parameters$pev_profiles, 'dl')),
+        parameters
+      )
+      vmax <- vnapply(parameters$pev_profiles, function(p) p$vmax)
+      beta <- vnapply(parameters$pev_profiles, function(p) p$beta)
+      alpha <- vnapply(parameters$pev_profiles, function(p) p$alpha)
+      vaccine_efficacy[vaccinated] <- calculate_pev_efficacy(
+        antibodies,
+        vmax[pev_profile],
+        beta[pev_profile],
+        alpha[pev_profile]
+      )
+    }
+    
+    # infection_rates <- rep(0, length = parameters$human_population)
+    if(parameters$parasite == "falciparum"){
+      
+      ## p.f models blood immunity
+      prob <- blood_immunity(variables$ib$get_values(source_humans), parameters)
+      
+    } else if (parameters$parasite == "vivax"){
+      
+      ## p.v does not model blood immunity but must take into account multiple bites per person
+      b <- 1 - (1 - parameters$b)^n_bites_per_person[bitten_humans$to_vector()]
+      
+      ## get infection rates for all bitten or with hypnozoites
+      infection_rates <- rep(0, length = source_humans$size())
+      infection_rates[bitset_index(source_humans, bitten_humans)] <- prob_to_rate(b)
+      
+      # Add relapse rates for individuals with hypnozoites
+      if(hypnozoites_humans$size()>0){
+        relapse_rates <- variables$hypnozoites$get_values(hypnozoites_humans) * parameters$f
+        hyp_source_index <- bitset_index(source_humans, hypnozoites_humans)
+        infection_rates[hyp_source_index] <- infection_rates[hyp_source_index] + relapse_rates
+        # Get and store relative rates for bite/relapse competing hazards resolution
+        relative_rates <- relapse_rates/infection_rates[hyp_source_index]
+        infection_outcome$set_relative_rates(
+          hypnozoites_humans, 
+          relative_rates)
+      }
+      
+      prob <- rate_to_prob(infection_rates)
+    }
+    
+    prob <- prob * (1 - prophylaxis) * (1 - vaccine_efficacy)
+    infection_rates <- prob_to_rate(prob)
+    
+    ## probability of incidence must be rendered at each timestep
+    incidence_probability_renderer(
+      variables$birth,
+      renderer,
+      source_humans,
+      prob,
+      "inc_",
+      parameters$incidence_min_ages,
+      parameters$incidence_max_ages,
+      timestep
     )
+    
+    ## capture infection rates to resolve in competing hazards
+    infection_outcome$set_rates(
+      source_humans, 
+      infection_rates)
```

Comment:
The edit to CompetingOutcome would allow you to add prob and relative_rates here...